doc(fivetran_sdk): Updated development guide for partners (#43)
fivetran/fivetran_sdk: development-guide.md

- Encrypt HTTP requests: Things like URLs, URL parameters, and query params are always encrypted for logging, and customer approval is needed to decrypt and examine them.

## Setup Form Guidelines
- Keep the form clear and concise, only requesting essential information for successful connector setup.
- Use clear and descriptive labels for each form field. Avoid technical jargon if possible.
- Organize the fields in a logical order that reflects the setup process.

### RPC Calls
#### ConfigurationForm

> This operation retrieves all the setup form fields and tests info required to establish a connection. You can provide various parameters for the fields to enhance the user experience, such as descriptions, optional fields, and more.

Author: > Is this statement needed - `tests info required to establish a connection`.

The configurationForm retrieves all the setup form fields and tests. 
Should I just change the sentence to `This operation retrieves all the setup form fields and tests` ?

> This operation retrieves all the setup form fields and tests info required to establish a connection. You can provide various parameters for the fields to enhance the user experience, such as descriptions, optional fields, and more.

Author: I have changed the statement, please take a look.

#### Test
The previous RPC call retrieves the tests that need to be executed during connection setup. This operation then invokes the test with the customer's credentials as parameters. Finally, it should return a success or failure indication for the test execution.

## Source Connector Guidelines

- Don't push anything other than source data to the destination. State will be saved to production DB and returned in `UpdateRequest`.
- Don't forget to handle new schemas/tables/columns per the information and user choices in `UpdateRequest#selection`.
- Make sure you checkpoint at least once an hour. The more frequently you do it, the better.

### RPC Calls
#### Schema

> This operation should retrieve all the information about the customer's schemas, tables, and columns. It also offers an optional `selection_not_supported` field that indicates whether customers can select or deselect tables and columns within the Fivetran dashboard.

Reviewer: This operation retrieves of customer's schemas, tables, and columns. It also includes an optional `selection_not_supported` field that indicates whether customers can select or deselect tables and columns within the Fivetran dashboard.

#### Update

> This operation should retrieve data from the source. We provide a request using the `UpdateRequest` message, which includes the customer's state, credentials, and schema information. The response, delivered through the `UpdateResponse` message, should contain data records or other supported operations.

Reviewer 2: ```suggestion
This operation should retrieve data from the source. We send a request using the `UpdateRequest` message, which includes the customer's state, credentials, and schema information. The response, streaming through the `UpdateResponse` message, can contain data records and other supported operations.
```

- This operation might be requested for a table that does not exist in the destination. In that case, it should NOT fail, simply ignore the request and return `success = true`.
- `utc_delete_before` has millisecond precision.

#### WriteBatch

> This operation provides details about the batch files containing the retrieved data. We provide a parameter `WriteBatchRequest` which contains all the information required for you to read the batch files. Here are some of the fields included in the request message:

Reviewer 3: ```suggestion
This operation provides details about the batch files containing the retrieved data. We provide the `WriteBatchRequest` parameter that contains all the information required for you to read the batch files. Here are some of the fields included in the request message:
```

> This operations provides details about the batch files containing the retrieved data. We provide a parameter `WriteBatchRequest` which contains all the information required for you to read the batch files. Here are some of the fields included in the request message::

Reviewer 3: ```suggestion
This operation provides details about the batch files containing the retrieved data. We provide a parameter `WriteBatchRequest` which contains all the information required for you to read the batch files. Here are some of the fields included in the request message:
```

> - `replace_files` is for `upsert` operation where the rows should be inserted if they don't exist or updated if they do. Each row will always provide values for all columns. Set the `_fivetran_synced` column in the destination with the values coming in from the csv files.

Reviewer 3: ```suggestion
- `replace_files` is for the `upsert` operation where the rows should be inserted if they don't exist or updated if they do. Each row will always provide values for all columns. Set the `_fivetran_synced` column in the destination with the values coming in from the .csv files.
```

> This operation provides details about the batch files containing the retrieved data. We provide the `WriteBatchRequest` parameter that contains all the information required for you to read the batch files. Here are some of the fields included in the request message:

Reviewer 2: ```suggestion
This operation provides details about the batch files containing the records to be pushed to the destination. We provide the `WriteBatchRequest` parameter that contains all the information required for you to read the batch files. Here are some of the fields included in the request message:
```

> - `update_files` is for `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the csv files.

Reviewer 3: ```suggestion
- `update_files` is for the `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the .CSV files.
```

> - `update_files` is for the `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `FileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the batch files.

Author: Note: I will merging the `CsvFileParams` and `ParquetFileParams` into one i.e. `FileParams`

> - `update_files` is for the `update` operation where modified columns have actual values whereas unmodified columns have the special value `unmodified_string` in `CsvFileParams`. Soft-deleted rows will arrive in here as well. Update the `_fivetran_synced` column in the destination with the values coming in from the .CSV files.

Reviewer 4: Now partners can also get parquet file as well, so this .CSV files doesn't make sense

Reviewer 2: Good catch @fivetran-abdulsalam! We should change `CsvFileParams` to `file_params` which covers both CSV and Parquet

- `delete_files` is for `hard delete` operation. Use primary key columns (or `_fivetran_id` system column for primary-keyless tables) to perform `DELETE FROM`.

> - `keys` is a map that provides a list of secret keys that can be used to decrypt batch files.

Reviewer 2: ```suggestion
- `keys` is a map that provides a list of secret keys, one for each batch file, that can be used to decrypt them.
```

- `file_params` provides information about the file type and any configurations applied to it, such as encryption or compression.

Also, Fivetran will deduplicate operations such that each primary key will show up only once in any of the operations

> Do not assume order of columns in the batch files. Always read the CSV file header to determine column order.

Reviewer 4: ```suggestion
Do not assume order of columns in the batch files. Always read the columns of batch file to determine column order.
```

    - `null_string` value is used to represent `NULL` value in all batch files.
    - `unmodified_string` value is used to indicate columns in `update_files` where the values did not change.

#### Capabilities

> This operation offers several additional features, as listed below:

Reviewer 2: ```suggestion
This operation offers the ability for the partner code to declare its choices for capabilities listed below:
```

> - Datatype Mappings: Supports adjusting partner data types for each Fivetran data type.

Reviewer 2: ```suggestion
- Datatype Mappings: Provides the option to map destination data types to Fivetran data types.
```

- Max value for columns: Provides an option to specify the maximum value for each data type.

#### AlterTable

> This operation is used to communicate updates to a table. The `SchemaDiff` message within the `AlterTableRequest` parameter provides details about these updates:

Reviewer 3: ```suggestion
This operation is used to communicate updates to a table. The `SchemaDiff` message within the `AlterTableRequest` parameter provides the details about these updates:
```

> This operation is used to communicate updates to a table. The `SchemaDiff` message within the `AlterTableRequest` parameter provides the details about these updates:

Reviewer 2: ```suggestion
This operation is used to communicate changes to a table as specific update operations. The `SchemaDiff` message within the `AlterTableRequest` parameter provides the details:
```

> - Adding a column (`add_column`): Fivetran uses this field to provide information about a new column to be added in destination table.

Reviewer 3: ```suggestion
- Adding a column (`add_column`): Fivetran uses this field to provide information about a new column to be added in a destination table.
```

> - Update Column type (`change_column_type`): This field provides the info on updated Column from source that needs to be reflected in destination table.

Reviewer 3: ```suggestion
- Update Column type (`change_column_type`): This field provides information on updated type of a column in the source that needs to be reflected in a destination table.
```

> - Primary key updates (`new_primary_key`):  If the customer's primary key have changed, this field lists all the new primary key used.

Author: @5tran-alexil What about this?

`Primary key updates (`new_primary_key`):  If the primary key has changed, this field lists all the new columns used in the revised primary key.`

> - Primary key updates (`new_primary_key`):  If the customer's primary key have changed, this field lists all the new primary key used.

Reviewer 3: ```suggestion
- Primary key updates (`new_primary_key`):  If the primary key have changed, this field lists all the new columns used in the updated primary key.
```

> - Primary key updates (`new_primary_key`):  If the customer's primary key have changed, this field lists all the new primary key used.

Reviewer 3: @fivetran-satvikpatil see ^^

> - Primary key updates (`new_primary_key`):  If the primary key has changed, this field lists all the new columns used in the updated primary key.

Reviewer 4: ```suggestion
- Primary key updates (`updated_primary_keys`):  If the primary key has changed, this field lists all the new columns used in the updated primary key.
```

Author: I have done the changes, please take a look.

> - Primary key updates (`updated_primary_keys`):  If the primary key has changed, this field lists all the columns used in the updated primary key.

Reviewer 4: We have changed it to `update_primary_keys`
```suggestion
- Primary key updates (`update_primary_keys`):  If the primary key has changed, this field lists all the columns used in the updated primary key.
```
